tokio/src/macros/try_join.rs

```diff
@@ -117,26 +120,55 @@ macro_rules! try_join {
         // the requirement of `Pin::new_unchecked` called below.
         let mut futures = ( $( maybe_done($e), )* );
 
+        // When poll_fn is polled, start polling the future at this index.
+        let mut start_index = 0;
+
         poll_fn(move |cx| {
+            const FUTURE_COUNT: u32 = $($total)*;
+
             let mut is_pending = false;
 
-            $(
-                // Extract the future for this branch from the tuple.
-                let ( $($skip,)* fut, .. ) = &mut futures;
+            let mut turn = start_index;
+
+            for _ in 0..FUTURE_COUNT {
+                $(
+                  {
+                      const INDEX: u32 = $($branch_index)*;
```

Comment:
This doesn't actually poll every future every time as far as I can tell, like the `loop` approach I suggested does.

Reply:
Maybe you could also add a test for this?

Reply:
Added a test, i think the futures are getting polled in the expected order. Let me know if i am missing something.

I can switch to the approach you suggested if you prefer.

Reply:
Ah, I misread it. Anyway, the current loop is quite wasteful because you go through every future with a smaller index every time you poll something.

Reply:
Got it.